openedx/features/content_type_gating/models.py

```diff
@@ -57,36 +50,24 @@ class ContentTypeGatingConfig(StackedConfigurationModel):
     )
 
     @classmethod
-    def has_full_access_role_in_masquerade(cls, user, course_key, course_masquerade, student_masquerade,
-                                           user_partition):
+    def has_full_access_role_in_masquerade(cls, user, course_key, course_masquerade):
         """
         The roles of the masquerade user are used to determine whether the content gate displays.
         The gate will not appear if the masquerade user has any of the following roles:
         Staff, Instructor, Beta Tester, Forum Community TA, Forum Group Moderator, Forum Moderator, Forum Administrator
         """
-        if student_masquerade:
+        if is_masquerading_as_specific_student(user, course_key):
             # If a request is masquerading as a specific user, the user variable will represent the correct user.
             if user and user.id and has_staff_roles(user, course_key):
                 return True
-        elif user_partition:
-            # If the current user is masquerading as a generic student in a specific group,
-            # then return the value based on that group.
-            masquerade_group = get_masquerading_user_group(course_key, user, user_partition)
-            if masquerade_group is None:
-                audit_mode_id = settings.COURSE_ENROLLMENT_MODES.get(CourseMode.AUDIT, {}).get('id')
-                # We are checking the user partition id here because currently content
-                # cannot have both the enrollment track partition and content gating partition
-                # configured simultaneously. We may change this in the future and allow
-                # configuring both partitions on content and selecting both partitions in masquerade.
-                if course_masquerade.user_partition_id == ENROLLMENT_TRACK_PARTITION_ID:
-                    return course_masquerade.group_id != audit_mode_id
-            elif masquerade_group is FULL_ACCESS:
-                return True
-            elif masquerade_group is LIMITED_ACCESS:
-                return False
+        elif (is_masquerading_as_staff(user, course_key) or
```

Comment:
Fair point, but this `has_full_access_role_in_masquerade` is only called once the calling code has already checked that we are masquerading. Maybe it's a reasonable refactor to move those checks lower into this method? But the calling code has other uses for knowing whether we're masquerading, so we wouldn't be cleaning much code up by moving this one check.

Reply:
I guess I look at it the other way around; if we're checking whether they're masquerading _every_ time we call this (1 for 1!) (`if course_masquerade:`), then we shouldn't leave it up to the calling function.

Granted, the code as-is works for now, but it's more brittle than it needs to be and is prone to non-obvious behavior.
eg, `is_masquerading_as_full_access` and related helpers accept a None-able `course_masquerade` argument. If anywhere else in the codebase starts calling `has_full_access_role_in_masquerade` (it is a public method, after all) without realizing that they need to first check if masquerading (it's not documented anywhere), then the codepath is unexpectedly different. Not only would the common case (non-staff learner, not masquerading) call each of our `is_masquerading_as_` helpers unsuccessfully in succession, but then most of these helper functions would _also_ call _additional_ sub-functions, as they interpret the lack of a `course_masquerade` object (either explicitly `None` or not passed at all) to mean they should first fetch the current masquerade settings.

If we weren't already modifying the logic of this function, I'd agree it's over-optimization. But since our changes here would _introduce_ this new edge case, I'd prefer if we localize the dependent logic and perform the check here.

So in practical terms, I think that means returning early at the beginning of this method:

```py
if not course_masquerade:
    return False
```

and then remove the `if course_masquerade:` check from `enabled_for_enrollment`.

Reply:
My changes don't introduce the edge case? This method always required masquerading to be checked above it.

Reply:
Perhaps by "introduced this edge case", what I meant was, "manifests new behavior when this edge case is triggered".

The issue isn't that "this method could be called with `course_masquerade=None`;
it's that "calling this with `course_masquerade=None` will add a half-dozen or so function calls to the call stack on `>99%` of all calls".

That's what I meant was introduced.

The calling method, `enabled_for_enrollment`, did always check for masquerading prior to calling `is_masquerading_as_full_access`.

The issue is that it's a public method so it could be called from anywhere else; we shouldn't assume they'll make an undocumented check prior to calling it. So you _could_ go ahead and document it, so it's at least a little more possible for people to know about this pre-condition, but I don't think that's much more helpful, especially when we can fix it by moving an `if` check :shrug: